File: Sources/Xarita/UI/CallTreeView.swift
import SwiftUI

/// A focused, readable view of one declaration's immediate neighbourhood.
///
/// This replaces the whole-codebase graph that preceded it. A force-directed
/// layout of a real project produces a hairball: technically correct, visually
/// impressive, and unreadable, because no one can follow sixteen hundred nodes.
/// Here the view never shows more than eleven — five callers, the subject, five
/// callees — laid out in fixed columns so the eye always knows where to look.
struct CallTreeView: View {

    @EnvironmentObject private var state: AppState
    @EnvironmentObject private var loc: Localization

    private let rowHeight: CGFloat = 30
    private let columnWidth: CGFloat = 176
    private let maxPerSide = 5

    var body: some View {
        if let graph = state.graph,
           let id = state.selection,
           id < graph.nodes.count {

            let node = graph.nodes[id]
            let (allCallers, allCallees) = graph.neighbours(of: id)
            let callers = Array(allCallers.prefix(maxPerSide))
            let callees = Array(allCallees.prefix(maxPerSide))
            let rows = max(1, max(callers.count, callees.count))
            let height = CGFloat(rows) * rowHeight + 34

            ZStack {
                connectors(callers: callers, callees: callees, rows: rows)
                columns(node: node, graph: graph,
                        callers: callers, callees: callees,
                        hiddenCallers: allCallers.count - callers.count,
                        hiddenCallees: allCallees.count - callees.count,
                        rows: rows)
            }
            .frame(height: height)
            .frame(maxWidth: .infinity)
            .padding(.vertical, 10)
            .background(Theme.surface)
        }
    }

    // MARK: - Geometry

    /// Vertical centre of row `index` within a column holding `count` rows,
    /// centred against the tallest column.
    private func y(index: Int, count: Int, rows: Int) -> CGFloat {
        let blockHeight = CGFloat(count) * rowHeight
        let totalHeight = CGFloat(rows) * rowHeight
        let top = (totalHeight - blockHeight) / 2
        return top + CGFloat(index) * rowHeight + rowHeight / 2
    }

    private func connectors(callers: [Int], callees: [Int], rows: Int) -> some View {
        Canvas { context, size in
            let midX = size.width / 2
            let leftEdge = midX - columnWidth / 2 - 8
            let rightEdge = midX + columnWidth / 2 + 8
            let centreY = CGFloat(rows) * rowHeight / 2 + 17
            let sourceX = midX - columnWidth / 2
            let targetX = midX + columnWidth / 2

            for (i, _) in callers.enumerated() {
                let startY = y(index: i, count: callers.count, rows: rows) + 17
                var path = Path()
                path.move(to: CGPoint(x: leftEdge, y: startY))
                path.addCurve(to: CGPoint(x: sourceX, y: centreY),
                              control1: CGPoint(x: leftEdge + 34, y: startY),
                              control2: CGPoint(x: sourceX - 34, y: centreY))
                context.stroke(path, with: .color(Theme.edgeIncoming.opacity(0.55)), lineWidth: 1.3)
            }

            for (i, _) in callees.enumerated() {
                let endY = y(index: i, count: callees.count, rows: rows) + 17
                var path = Path()
                path.move(to: CGPoint(x: targetX, y: centreY))
                path.addCurve(to: CGPoint(x: rightEdge, y: endY),
                              control1: CGPoint(x: targetX + 34, y: centreY),
                              control2: CGPoint(x: rightEdge - 34, y: endY))
                context.stroke(path, with: .color(Theme.edgeOutgoing.opacity(0.55)), lineWidth: 1.3)
            }
        }
    }

    // MARK: - Columns

    private func columns(node: GraphNode, graph: CodeGraph,
                         callers: [Int], callees: [Int],
                         hiddenCallers: Int, hiddenCallees: Int,
                         rows: Int) -> some View {
        HStack(spacing: 16) {
            column(ids: callers, hidden: hiddenCallers, rows: rows,
                   title: loc.t.callers, tint: Theme.edgeIncoming,
                   empty: loc.t.noCallers, alignment: .trailing)

            CentreChip(node: node)
                .frame(width: columnWidth)

            column(ids: callees, hidden: hiddenCallees, rows: rows,
                   title: loc.t.callees, tint: Theme.edgeOutgoing,
                   empty: loc.t.noCallees, alignment: .leading)
        }
    }

    private func column(ids: [Int], hidden: Int, rows: Int,
                        title: String, tint: Color, empty: String,
                        alignment: HorizontalAlignment) -> some View {
        VStack(alignment: alignment, spacing: 0) {
            Text(title.uppercased())
                .font(Theme.Font.micro)
                .tracking(0.6)
                .foregroundStyle(tint.opacity(0.9))
                .frame(height: 17)

            if ids.isEmpty {
                Text(empty)
                    .font(Theme.Font.micro)
                    .foregroundStyle(Theme.textTertiary)
                    .frame(height: CGFloat(rows) * rowHeight, alignment: .center)
            } else {
                VStack(spacing: 0) {
                    ForEach(ids, id: \.self) { id in
                        NeighbourChip(id: id, alignment: alignment)
                            .frame(height: rowHeight)
                    }
                    if hidden > 0 {
                        Text("+\(hidden)")
                            .font(Theme.Font.micro)
                            .foregroundStyle(Theme.textTertiary)
                    }
                }
                .frame(height: CGFloat(rows) * rowHeight,
                       alignment: ids.count == rows ? .top : .center)
            }
        }
        .frame(width: columnWidth)
    }
}

/// The subject of the current view.
private struct CentreChip: View {
    let node: GraphNode

    var body: some View {
        VStack(spacing: 3) {
            Text(node.name)
                .font(Theme.Font.mono.weight(.semibold))
                .foregroundStyle(Theme.textPrimary)
                .lineLimit(1)
                .truncationMode(.middle)
            if let container = node.container {
                Text(container)
                    .font(Theme.Font.micro)
                    .foregroundStyle(Theme.textTertiary)
                    .lineLimit(1)
            }
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 9)
        .frame(maxWidth: .infinity)
        .background(
            RoundedRectangle(cornerRadius: 8)
                .fill(Theme.accentMuted)
        )
        .overlay(
            RoundedRectangle(cornerRadius: 8)
                .strokeBorder(Theme.accent, lineWidth: 1.4)
        )
    }
}

private struct NeighbourChip: View {
    @EnvironmentObject private var state: AppState
    let id: Int
    let alignment: HorizontalAlignment
    @State private var hovering = false

    var body: some View {
        if let graph = state.graph, id < graph.nodes.count {
            let node = graph.nodes[id]
            Button {
                state.select(id)
            } label: {
                HStack(spacing: 6) {
                    if alignment == .leading {
                        dot(node)
                        label(node)
                    } else {
                        label(node)
                        dot(node)
                    }
                }
                .padding(.horizontal, 9)
                .padding(.vertical, 5)
                .frame(maxWidth: .infinity, alignment: alignment == .leading ? .leading : .trailing)
                .background(
                    RoundedRectangle(cornerRadius: 6)
                        .fill(hovering ? Theme.accentMuted : Theme.surfaceRaised)
                )
                .overlay(
                    RoundedRectangle(cornerRadius: 6)
                        .strokeBorder(Theme.border, lineWidth: 1)
                )
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .onHover { hovering = $0 }
        }
    }

    private func dot(_ node: GraphNode) -> some View {
        Circle()
            .fill(Theme.color(for: node.kind, external: node.isExternal))
            .frame(width: 6, height: 6)
    }

    private func label(_ node: GraphNode) -> some View {
        Text(node.name)
            .font(Theme.Font.monoSmall)
            .foregroundStyle(Theme.textPrimary)
            .lineLimit(1)
            .truncationMode(.middle)
    }
}

File: Windows/Atlas.Windows/Views/CallTreeView.cs
using System.Globalization;
using Avalonia;
using Avalonia.Controls;
using Avalonia.Media;
using Atlas.Windows.Engine;

namespace Atlas.Windows.Views;

/// <summary>
/// One declaration's immediate neighbourhood: who calls it, what it calls.
///
/// Never more than eleven things — five callers, the subject, five callees —
/// in fixed columns, so the eye always knows where to look. The whole-codebase
/// graph this replaced produced a hairball: technically correct, visually
/// impressive, and unreadable, because nobody can follow sixteen hundred nodes.
///
/// Ported from Sources/Atlas/UI/CallTreeView.swift.
/// </summary>
public sealed class CallTreeView : Control
{
    private const double RowHeight = 30;
    private const double ColumnWidth = 176;
    private const int MaxPerSide = 5;

    private readonly Report _report;
    private readonly Strings _t;
    private readonly SymbolEntry? _subject;
    private readonly List<int> _callers = [];
    private readonly List<int> _callees = [];
    private readonly int _hiddenCallers;
    private readonly int _hiddenCallees;
    private readonly int _rows;

    public CallTreeView(Report report, Strings t, int index)
    {
        _report = report;
        _t = t;

        if (index >= 0 && index < report.Symbols.Count)
        {
            _subject = report.Symbols[index];
            var allCallers = report.Calls.Where(c => c.To == index)
                                   .Select(c => c.From).Distinct().ToList();
            var allCallees = report.Calls.Where(c => c.From == index)
                                   .Select(c => c.To).Distinct().ToList();
            _callers = allCallers.Take(MaxPerSide).ToList();
            _callees = allCallees.Take(MaxPerSide).ToList();
            _hiddenCallers = allCallers.Count - _callers.Count;
            _hiddenCallees = allCallees.Count - _callees.Count;
        }
        _rows = Math.Max(1, Math.Max(_callers.Count, _callees.Count));
    }

    protected override Size MeasureOverride(Size availableSize) =>
        new(availableSize.Width, _rows * RowHeight + 44);

    /// <summary>
    /// Vertical centre of row <paramref name="index"/> in a column of
    /// <paramref name="count"/>, centred against the tallest column.
    /// </summary>
    private double Y(int index, int count) =>
        17 + (_rows * RowHeight - count * RowHeight) / 2 + index * RowHeight + RowHeight / 2;

    public override void Render(DrawingContext context)
    {
        if (_subject is null) return;
        var size = Bounds.Size;
        if (size.Width <= 0) return;

        context.FillRectangle(Broadsheet.Brush(Broadsheet.Surface),
                              new Rect(0, 0, size.Width, size.Height));

        double midX = size.Width / 2;
        double leftEdge = midX - ColumnWidth / 2 - 8;
        double rightEdge = midX + ColumnWidth / 2 + 8;
        double centreY = _rows * RowHeight / 2 + 17;
        double sourceX = midX - ColumnWidth / 2;
        double targetX = midX + ColumnWidth / 2;

        // Incoming in magenta, outgoing in cyan. The same pairing as the Map
        // and the inspector, doing the same job: pointing a direction.
        for (int i = 0; i < _callers.Count; i++)
        {
            Curve(context, new Point(leftEdge, Y(i, _callers.Count)),
                  new Point(sourceX, centreY), 34,
                  Broadsheet.Fade(Broadsheet.EdgeIncoming, 0.55));
        }
        for (int i = 0; i < _callees.Count; i++)
        {
            Curve(context, new Point(targetX, centreY),
                  new Point(rightEdge, Y(i, _callees.Count)), 34,
                  Broadsheet.Fade(Broadsheet.EdgeOutgoing, 0.55));
        }

        // The side columns start clear of the centre chip, which is
        // ColumnWidth wide and centred on midX — not clear of midX itself, or
        // the names print over the subject.
        Column(context, _callers, _hiddenCallers, leftEdge - 16 - ColumnWidth,
               _t["callers"], Broadsheet.EdgeIncoming, _t["noCallers"], rightAligned: true);
        Column(context, _callees, _hiddenCallees, rightEdge + 16,
               _t["callees"], Broadsheet.EdgeOutgoing, _t["noCallees"], rightAligned: false);
        Centre(context, midX, centreY);
    }

    private static void Curve(DrawingContext context, Point from, Point to,
                              double bend, Color ink)
    {
        var geometry = new StreamGeometry();
        using (var pen = geometry.Open())
        {
            pen.BeginFigure(from, false);
            pen.CubicBezierTo(new Point(from.X + bend, from.Y),
                              new Point(to.X - bend, to.Y), to);
            pen.EndFigure(false);
        }
        context.DrawGeometry(null, new Pen(Broadsheet.Brush(ink), 1.3), geometry);
    }

    private void Column(DrawingContext context, List<int> ids, int hidden, double x,
                        string title, Color tint, string empty, bool rightAligned)
    {
        var heading = Text(title.ToUpperInvariant(), Broadsheet.Fonts.Serif,
                           Broadsheet.Fonts.Micro, Broadsheet.Fade(tint, 0.9));
        context.DrawText(heading,
            new Point(rightAligned ? x + ColumnWidth - heading.Width : x, 0));

        if (ids.Count == 0)
        {
            var none = Text(empty, Broadsheet.Fonts.Serif,
                            Broadsheet.Fonts.Micro, Broadsheet.TextTertiary);
            context.DrawText(none, new Point(
                rightAligned ? x + ColumnWidth - none.Width : x,
                17 + _rows * RowHeight / 2 - none.Height / 2));
            return;
        }

        for (int i = 0; i < ids.Count; i++)
        {
            var symbol = _report.Symbols[ids[i]];
            var name = Text(symbol.Display, Broadsheet.Fonts.Mono,
                            Broadsheet.Fonts.MonoSmall, Broadsheet.TextSecondary);
            name.MaxTextWidth = ColumnWidth;
            name.MaxLineCount = 1;
            name.Trimming = TextTrimming.CharacterEllipsis;
            context.DrawText(name, new Point(
                rightAligned ? x + ColumnWidth - name.Width : x,
                Y(i, ids.Count) - name.Height / 2));
        }

        if (hidden > 0)
        {
            var more = Text($"+{hidden.ToString(CultureInfo.InvariantCulture)}",
                            Broadsheet.Fonts.Serif, Broadsheet.Fonts.Micro,
                            Broadsheet.TextTertiary);
            context.DrawText(more, new Point(
                rightAligned ? x + ColumnWidth - more.Width : x,
                Y(ids.Count - 1, ids.Count) + RowHeight / 2));
        }
    }

    private void Centre(DrawingContext context, double midX, double centreY)
    {
        var name = Text(_subject!.Name, Broadsheet.Fonts.Mono, Broadsheet.Fonts.MonoSize,
                        Broadsheet.TextPrimary, FontWeight.SemiBold);
        name.MaxTextWidth = ColumnWidth;
        name.MaxLineCount = 1;
        name.Trimming = TextTrimming.CharacterEllipsis;

        double top = centreY - name.Height / 2 - (_subject.Container is null ? 0 : 8);
        var box = new Rect(midX - ColumnWidth / 2, top - 8, ColumnWidth,
                           name.Height + 16 + (_subject.Container is null ? 0 : 14));
        context.FillRectangle(Broadsheet.Brush(Broadsheet.SurfaceRaised), box);
        context.DrawRectangle(null, new Pen(Broadsheet.Brush(Broadsheet.Border), 1), box);

        context.DrawText(name, new Point(midX - name.Width / 2, top));

        if (_subject.Container is { Length: > 0 } container)
        {
            var owner = Text(container, Broadsheet.Fonts.Serif, Broadsheet.Fonts.Micro,
                             Broadsheet.TextTertiary);
            context.DrawText(owner, new Point(midX - owner.Width / 2, top + name.Height + 2));
        }
    }

    private static FormattedText Text(string value, FontFamily family, double size,
                                      Color ink, FontWeight weight = FontWeight.Normal) =>
        new(value, CultureInfo.CurrentCulture, FlowDirection.LeftToRight,
            new Typeface(family, FontStyle.Normal, weight), size, Broadsheet.Brush(ink));
}
